internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go

```diff
@@ -0,0 +1,255 @@
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	se := &intermediateTypes.ServiceEndpoints{
+		Resource: rsp.Resource,
+	}
+
+	var endpoints pbcatalog.ServiceEndpoints
+	err = rsp.Resource.Data.UnmarshalTo(&endpoints)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&endpoints, err)
+	}
+
+	se.Endpoints = &endpoints
+	return se, nil
+}
+
+func (f *Fetcher) FetchDestinations(ctx context.Context, id *pbresource.ID) (*intermediateTypes.Destinations, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	u := &intermediateTypes.Destinations{
+		Resource: rsp.Resource,
+	}
+
+	var destinations pbmesh.Upstreams
+	err = rsp.Resource.Data.UnmarshalTo(&destinations)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&destinations, err)
+	}
+
+	u.Destinations = &destinations
+	return u, nil
+}
+
+func (f *Fetcher) FetchDestinationsData(
+	ctx context.Context,
+	destinationRefs []intermediateTypes.CombinedDestinationRef,
+) ([]*intermediateTypes.Destination, map[string]*intermediateTypes.Status, error) {
+
+	var destinations []*intermediateTypes.Destination
+	statuses := make(map[string]*intermediateTypes.Status)
+	for _, dest := range destinationRefs {
+		// Fetch Destinations resource if there is one.
+		us, err := f.FetchDestinations(ctx, dest.ExplicitDestinationsID)
+		if err != nil {
+			// If there's an error, return and force another reconcile instead of computing
+			// partial proxy state.
+			return nil, statuses, err
+		}
+
+		if us == nil {
+			// If the Destinations resource is not found, then we should delete it from cache and continue.
+			f.Cache.DeleteDestination(dest.ServiceRef, dest.Port)
+			continue
+		}
+
+		d := &intermediateTypes.Destination{}
+		// As Destinations resource contains a list of destinations,
+		// we need to find the one that references our service and port.
+		d.Explicit = findDestination(dest.ServiceRef, dest.Port, us.Destinations)
+
+		// Fetch ServiceEndpoints.
+		serviceID := resource.IDFromReference(dest.ServiceRef)
+		se, err := f.FetchServiceEndpoints(ctx, resource.ReplaceType(catalog.ServiceEndpointsType, serviceID))
+		if err != nil {
+			return nil, statuses, err
+		}
+
+		serviceRef := resource.ReferenceToString(dest.ServiceRef)
+		upstreamsRef := resource.IDToString(us.Resource.Id)
+		if se == nil {
+			// If the Service Endpoints resource is not found, then we update the status of the Upstreams resource
+			// but don't remove it from cache in case it comes back.
+			updateStatusCondition(statuses, upstreamsRef, dest.ExplicitDestinationsID,
+				us.Resource.Status, us.Resource.Generation, ctrlStatus.ConditionDestinationServiceNotFound(serviceRef))
+			continue
```

Comment:
Is there any concern with early-terminating the loop before you generate any following status conditions?

Reply:
I don't think so. Once you've hit one error, I think it's ok not to generate other potential error conditions.